lib/semantics/resolve-names.cc

```diff
@@ -1099,7 +1207,62 @@ class OmpVisitor : public virtual DeclarationVisitor {
     return false;
   }
 
+  void Post(const parser::Name &);
+
 private:
+  struct OmpContext {
+    OmpContext(const parser::CharBlock &source, OmpDirective d)
+      : directiveSource{source}, directive{d} {}
+    parser::CharBlock directiveSource{nullptr};
+    OmpDirective directive;
+    Scope *scope{nullptr};
+    // TODO: default DSA is implicitly determined in different ways
+    Symbol::Flag defaultDSA{Symbol::Flag::OmpShared};
+    // variables on Data-sharing attribute clauses
+    std::map<const Symbol *, Symbol::Flag> objectMap;
+    bool withinConstruct{false};
+  };
+  // back() is the top of the stack
+  OmpContext &GetContext() {
+    CHECK(!ompContext_.empty());
+    return ompContext_.back();
+  }
+  void PushContext(const parser::CharBlock &source, OmpDirective dir) {
+    ompContext_.emplace_back(source, dir);
+    SetContextScope(source);
```

Comment:
If you made this `ompContext_.emplace_back(source, dir, context_.FindScope(source));` then `scope` in `OmpContext` could be a reference so you wouldn't have to worry about whether it is null. And you wouldn't need `SetContextScope`.

Reply:
Very cool!

Reply:
I've changed the `scope` to reference.